### ISSUE TITLE
IDFileConverter: Unexpected internal error (Could not convert string '' to a double value)

### ISSUE THREAD

**Issue by elrubio** [OpenMS/OpenMS#7097]
I have a set of *mzIdentML* files (exported from Mascot) that pass validation (some warnings, no errors), but `IDFileConverter` throws an `Unexpected internal error` when trying to convert them to e.g. *idXML*:

```
$ IDFileConverter -in /data/F003285.mzid -out /data/F003285.idXML -debug 9
[...]
TOPPBase.cpp(1605): Value of string option 'no_progress': 0
TOPPBase.cpp(1605): Value of string option 'in': /data/F003285.mzid
TOPPBase.cpp(1605): Checking input file '/data/F003285.mzid'
TOPPBase.cpp(1605): Value of string option 'mz_file':
TOPPBase.cpp(1605): Value of string option 'out': /data/F003285.idXML
TOPPBase.cpp(1605): Checking output file '/data/F003285.idXML'
TOPPBase.cpp(1605): Value of string option 'out_type':
Progress of 'Loading...':
Converting from mzid: you might experience loss of information depending on the capabilities of the target format.
No DatabaseName element found, use read in results at own risk.
Error: Unexpected internal error (Could not convert string '' to a double value)
TOPPBase.cpp(1605): Error occurred in line 139 of file /opt/conda/conda-bld/openms-meta_1688467064501/work/src/openms/include/OpenMS/DATASTRUCTURES/StringUtils.h (in function: static double OpenMS::StringUtilsHelper::toDouble(const OpenMS::String&)) !
<No DatabaseName element found, use read in results at own risk.> occurred 2 times
```

Tested versions: 
- Version: 2.9.1-pre-exported-20230704 Jul  4 2023, 10:44:31 (docker)
- Version: 3.0.0 Jul 14 2023, 11:10:53, Revision: be787e9

Additional questions:
1. Is there a way to enable a stack trace in the error output?
2. Is there a way to enable output of the offending location in the file (line number or element XPath)?

Example file:
- [F003285.mzid.zip](https://github.com/OpenMS/OpenMS/files/12701260/F003285.mzid.zip)

<details>

<summary>Validator output</summary>

Validator used: https://github.com/HUPO-PSI/mzIdentML/releases/download/mzIdentMLValidator_GUI_v1.4.35-SNAPSHOT/mzIdentMLValidator_GUI_v1.4.35-SNAPSHOT.zip

```
The following messages were obtained during the validation of your XML file:


Message 1:
    Level: WARN
    --> unanticipated terms for XPath '/MzIdentML/AnalysisProtocolCollection/ProteinDetectionProtocol/analysisParams/cvParam/@accession' : [MS:1003030]


Message 2:
    Rule ID: AuditCollectionPerson_should_rule
    Level: WARN
    Context(/person/cvParam/@accession ) in 2 locations
    --> None of the given CvTerms were found at '/MzIdentML/AuditCollection/person/cvParam/@accession' because no values were found:
  - The sole term MS:1000587 (contact address) or any of its children. The term can be repeated. The matching value has to be the identifier of the term, not its name.
  - The sole term MS:1000588 (contact URL) or any of its children. The term can be repeated. The matching value has to be the identifier of the term, not its name.
  - The sole term MS:1000589 (contact email) or any of its children. The term can be repeated. The matching value has to be the identifier of the term, not its name.
  - The sole term MS:1000590 (contact affiliation) or any of its children. The term can be repeated. The matching value has to be the identifier of the term, not its name.
  - The sole term MS:1001755 (contact phone number) or any of its children. The term can be repeated. The matching value has to be the identifier of the term, not its name.
  - The sole term MS:1001756 (contact fax number) or any of its children. The term can be repeated. The matching value has to be the identifier of the term, not its name.
  - The sole term MS:1001757 (contact toll-free phone number) or any of its children. The term can be repeated. The matching value has to be the identifier of the term, not its name.


Message 3:
    Rule ID: AuditCollectionOrganization_should_rule
    Level: WARN
    Context(/organization/cvParam/@accession ) in 2 locations
    --> None of the given CvTerms were found at '/MzIdentML/AuditCollection/organization/cvParam/@accession' because no values were found:
  - The sole term MS:1000586 (contact name) or any of its children. The term can be repeated. The matching value has to be the identifier of the term, not its name.
  - The sole term MS:1000587 (contact address) or any of its children. The term can be repeated. The matching value has to be the identifier of the term, not its name.
  - The sole term MS:1000588 (contact URL) or any of its children. The term can be repeated. The matching value has to be the identifier of the term, not its name.
  - The sole term MS:1000589 (contact email) or any of its children. The term can be repeated. The matching value has to be the identifier of the term, not its name.
  - The sole term MS:1001755 (contact phone number) or any of its children. The term can be repeated. The matching value has to be the identifier of the term, not its name.
  - The sole term MS:1001756 (contact fax number) or any of its children. The term can be repeated. The matching value has to be the identifier of the term, not its name.
  - The sole term MS:1001757 (contact toll-free phone number) or any of its children. The term can be repeated. The matching value has to be the identifier of the term, not its name.


Message 4:
    Rule ID: SpectrumIdentificationList_may_rule
    Level: INFO
    Context(/cvParam/@accession ) in 2 locations
    --> None of the given CvTerms were found at '/MzIdentML/DataCollection/AnalysisData/SpectrumIdentificationList/cvParam/@accession' because no values were found:
  - Any children term of MS:1001184 (search statistics). The term can be repeated. The matching value has to be the identifier of the term, not its name.


Message 5:
    Rule ID: SearchDatabaseDatabaseName_may_rule
    Level: INFO
    Context(/searchDatabase/databaseName/cvParam/@accession ) in 4 locations
    --> None of the given CvTerms were found at '/MzIdentML/DataCollection/Inputs/searchDatabase/databaseName/cvParam/@accession' because no values were found:
  - Any children term of MS:1001013 (database name). A single instance of this term can be specified. The matching value has to be the identifier of the term, not its name.


Message 6:
    Rule ID: MassTable_may_rule
    Level: INFO
    Context(/massTable/cvParam/@accession ) in 2 locations
    --> None of the given CvTerms were found at '/MzIdentML/AnalysisProtocolCollection/SpectrumIdentificationProtocol/massTable/cvParam/@accession' because no values were found:
  - Any children term of MS:1001354 (mass table options). The term can be repeated. The matching value has to be the identifier of the term, not its name.


Message 7:
    Rule ID: CvListObjectRule
    Level: INFO
    Context(/MzIdentML/cvList )
    --> The cv element for UO uses an old URI.
    Tip: Provide the newest version for all CV elements under the CvList element.


Message 8:
    Rule ID: ProteinAmbiguityGroup_may_rule
    Level: INFO
    Context(/MzIdentML/DataCollection/AnalysisData/ProteinDetectionList/proteinAmbiguityGroup/cvParam/@accession )
    --> The term 'protein ambiguity group result details' (MS:1001147) doesn't have children in the ontology 'MS', yet the CvMapping recommends to use them. The CvTerm will be removed.


Message 9:
    Rule ID: SourceFile_may_rule
    Level: INFO
    Context(/sourceFile/cvParam/@accession ) in 2 locations
    --> None of the given CvTerms were found at '/MzIdentML/DataCollection/Inputs/sourceFile/cvParam/@accession' because no values were found:
  - Any children term of MS:1000561 (data file checksum type). The term can be repeated. The matching value has to be the identifier of the term, not its name.


======== RULE STATISTICS ========
Invalid XML schema validation: 0

CvMappingRule total count: 50
CvMappingRules not run: 3
CvMappingRules run & not matching: 7
CvMappingRules invalid XPath: 2
CvMappingRules run & valid: 39

ObjectRules total count: 21
ObjectRules not run: 2
ObjectRules run & not matching: 1
ObjectRules run & valid: 18

Unanticipated CV terms: 1
Not matching messages received: 7
Messages not reported since they occur more than 1 times: 2

```

</details>

**Comment by jpfeuffer** [comment]
Hi!

No, unfortunately you cannot get more information without using a C++ Debugger.

You can try to remove the DatabaseName elements since I found a bug in our parser when handling them. Not sure if this was the real cause of the error though.

**Comment by elrubio** [comment]
Thanks for the fix, but it's indeed unrelated. I have the same issue with the latest nightly (that includes your fix). 

I spotted [another bug](7099) though concerning the `DatabaseName` warning.

**Comment by elrubio** [comment]
I eventually took the time to setup and run the debugger. The culprit was easy to find:

The mzIdentML v1.1 file produced by Mascot contains the following [valid](https://peptideatlas.org/tmp/mzIdentML/mzIdentML1.1.0.html#PeptideEvidenceRef) cvParam elements. Note that the last one doesn't have a *value* attribute:

```
<SpectrumIdentificationResult id="SIR_66" spectrumID="index=6182" spectraData_ref="SD_1">
  <SpectrumIdentificationItem id="SII_66_1"  calculatedMassToCharge="371.732122" chargeState="2" experimentalMassToCharge="371.731686" peptide_ref="peptide_66_1" rank="1" passThreshold="false">
    <PeptideEvidenceRef peptideEvidence_ref="PE_66_1_1_P62911|RL32_MOUSE_0_115_121" />
    <cvParam accession="MS:1001171" name="Mascot:score" cvRef="PSI-MS" value="20.18" />
    <cvParam accession="MS:1001172" name="Mascot:expectation value" cvRef="PSI-MS" value="0.22" />
    <cvParam accession="MS:1001363" name="peptide unique to one protein" cvRef="PSI-MS" />
  </SpectrumIdentificationItem>
  [...]
</SpectrumIdentificationResult>
``` 

The parser only checks for and correctly handles two specific CV terms, but blatantly assumes that all other terms a) have a value and b) that the value is numeric. This assumption is not even guarded, so it chokes and dies.

https://github.com/OpenMS/OpenMS/blob/55094ca9c98949c8d8657d233267a56978f3de8d/src/openms/source/FORMAT/HANDLERS/MzIdentMLDOMHandler.cpp#L2084-L2103

**Comment by cbielow** [comment]
fix is coming up. I just want to test a full Mascot mzIdentML, just to make sure there is nothing else in these files which is problematic.

**Comment by cbielow** [comment]
fixed in #7153
